src/Raven.Pal/src/posix/getsysteminformation.c

```diff
@@ -48,3 +63,35 @@ rvn_get_path_disk_space(const char* path, uint64_t* total_free_bytes, uint64_t*
 
     return SUCCESS;
 }
+
+EXPORT int32_t
+rvn_get_path_disk_stats(const char * path, struct IO_STATS* io_stats, int32_t* detailed_error_code)
```

Comment:
wouldn't it be better to move that to C#? We already have smaps readers there, so I do not see a problem to have this there as well. All Pal code comes with high maintainability cost IMO.

Reply:
Talked with Igal. He mentioned he expected it to be more closer to the system in the beginning, but it turns out we just need `stat`, `major` and `minor`. The first one is easy to have in C#. The latter 2 are just C macros:
```
#define MINORBITS        20
#define MINORMASK        ((1U << MINORBITS) - 1)

#define MAJOR(dev)        ((unsigned int) ((dev) >> MINORBITS))
#define MINOR(dev)        ((unsigned int) ((dev) & MINORMASK))
#define MKDEV(ma,mi)      (((ma) << MINORBITS) | (mi))
```

https://stackoverflow.com/questions/20640043/major-and-minor-function-in-linux-kernel

Reply:
@ppekrol, this is true for the whole Pal library, no?
As I understand we want to encapsulate the operating system part behind this library.
Of course, I can write the same code in C# and call the `stat` as I do for the Pal function.

Reply:
They reason we have `PAL` is specifically to handle differences between OSes.
For example, Linux ARM vs. Linux x86 uses different syscall numbers.

Reply:
So if the difference is between Linux and Mac?
And for now, we have one syscall but later we may need more (maybe to calculate the sector size).

Reply:
No, right now, we don't care to implement this outside of Linux, and we don't need to worry about changes between systems.

Reply:
Sure, but that's just syscalls arch, so you're changing 32 to 64 in a few lines.  

The code above adds another closer-to-the-metal maintenance point, which we can avoid - smaps proc virt file is also parsed in C# and I don't think it's less performant to a significant point. As far as I recall we have ton of code that uses `PlatformDetails` and goes into different branches for different platforms. I got a feeling it's actually more about "it's done let's use it", but IMO I think it might make us cry later on and spend hours debugging - C is easier to crash, harder to maintain and the only advantage here is it's already written, but we don't know how much gray hair it might get us later?

In addition to that CPU and MEM stats are also pulled using C#.

Reply:
Also C# allows you to easily add tests for parsing and test for cases like those overflows mentioned before.